// @dcl/scene-runtime/dist/worker/client/ParcelIdentity.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParcelIdentityServiceClient = void 0;
const codegen = __importStar(require("@dcl/rpc/dist/codegen"));
const parcel_identity_gen_1 = require("@dcl/protocol/out-ts/decentraland/kernel/apis/parcel_identity.gen");
var ParcelIdentityServiceClient;
(function (ParcelIdentityServiceClient) {
    function create(clientPort) {
        return codegen.loadService(clientPort, parcel_identity_gen_1.ParcelIdentityServiceDefinition);
    }
    ParcelIdentityServiceClient.create = create;
    function createLegacy(clientPort) {
        const originalService = codegen.loadService(clientPort, parcel_identity_gen_1.ParcelIdentityServiceDefinition);
        return {
            ...originalService,
            /**
             * Returns the coordinates and the definition of a parcel
             */
            async getParcel() {
                const data = await originalService.getParcel({});
                return {
                    land: {
                        sceneId: data.land?.sceneId || '',
                        sceneJsonData: JSON.parse(data.land?.sceneJsonData || '{}'),
                        baseUrl: data.land?.baseUrl || '',
                        baseUrlBundles: data.land?.baseUrlBundles || '',
                        mappingsResponse: {
                            root_cid: data.land?.mappingsResponse?.rootCid || '',
                            parcel_id: data.land?.mappingsResponse?.parcelId || '',
                            contents: data.land?.mappingsResponse?.contents || []
                        }
                    },
                    cid: data.cid
                };
            },
            /**
             * Returns the scene id
             */
            async getSceneId() {
                const data = await originalService.getSceneId({});
                return data.sceneId;
            }
        };
    }
    ParcelIdentityServiceClient.createLegacy = createLegacy;
})(ParcelIdentityServiceClient = exports.ParcelIdentityServiceClient || (exports.ParcelIdentityServiceClient = {}));
