// @dcl/scene-runtime/dist/worker/client/RestrictedActions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestrictedActionsServiceClient = void 0;
const codegen = __importStar(require("@dcl/rpc/dist/codegen"));
const restricted_actions_gen_1 = require("@dcl/protocol/out-ts/decentraland/kernel/apis/restricted_actions.gen");
var RestrictedActionsServiceClient;
(function (RestrictedActionsServiceClient) {
    function create(clientPort) {
        return codegen.loadService(clientPort, restricted_actions_gen_1.RestrictedActionsServiceDefinition);
    }
    RestrictedActionsServiceClient.create = create;
    function createLegacy(clientPort) {
        const originalService = codegen.loadService(clientPort, restricted_actions_gen_1.RestrictedActionsServiceDefinition);
        return {
            ...originalService,
            /**
             * move player to a position inside the scene
             *
             * @param position PositionType
             * @param cameraTarget PositionType
             */
            async movePlayerTo(newPosition, cameraTarget) {
                await originalService.movePlayerTo({
                    newRelativePosition: newPosition,
                    cameraTarget: cameraTarget || undefined
                });
            },
            /**
             * trigger an emote on the current player
             *
             * @param emote the emote to perform
             */
            async triggerEmote(emote) {
                await originalService.triggerEmote({ predefinedEmote: emote.predefined });
            }
        };
    }
    RestrictedActionsServiceClient.createLegacy = createLegacy;
})(RestrictedActionsServiceClient = exports.RestrictedActionsServiceClient || (exports.RestrictedActionsServiceClient = {}));
