// @dcl/scene-runtime/dist/worker/client/SignedFetch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignedFetchServiceClient = void 0;
const codegen = __importStar(require("@dcl/rpc/dist/codegen"));
const signed_fetch_gen_1 = require("@dcl/protocol/out-ts/decentraland/kernel/apis/signed_fetch.gen");
var SignedFetchServiceClient;
(function (SignedFetchServiceClient) {
    function create(clientPort) {
        return codegen.loadService(clientPort, signed_fetch_gen_1.SignedFetchServiceDefinition);
    }
    SignedFetchServiceClient.create = create;
    function createLegacy(clientPort) {
        const originalService = codegen.loadService(clientPort, signed_fetch_gen_1.SignedFetchServiceDefinition);
        return {
            ...originalService,
            async signedFetch(url, originalInit) {
                let init = undefined;
                if (originalInit) {
                    init = { headers: {} };
                    if (originalInit.headers && typeof originalInit.headers === 'object') {
                        init.headers = originalInit.headers;
                    }
                    if (originalInit.body && typeof originalInit.body === 'string') {
                        init.body = originalInit.body;
                    }
                    if (originalInit.method && typeof originalInit.method === 'string') {
                        init.method = originalInit.method;
                    }
                }
                const responseBodyType = originalInit?.responseBodyType || 'text';
                const result = await originalService.signedFetch({ url, init });
                return {
                    ok: result.ok,
                    status: result.status,
                    statusText: result.statusText,
                    headers: result.headers,
                    json: responseBodyType === 'json' ? JSON.parse(result.body) : undefined,
                    text: responseBodyType === 'text' ? result.body : undefined
                };
            }
        };
    }
    SignedFetchServiceClient.createLegacy = createLegacy;
})(SignedFetchServiceClient = exports.SignedFetchServiceClient || (exports.SignedFetchServiceClient = {}));
