// @dcl/scene-runtime/dist/worker/client/UserIdentity.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserIdentityServiceClient = void 0;
const codegen = __importStar(require("@dcl/rpc/dist/codegen"));
const user_identity_gen_1 = require("@dcl/protocol/out-ts/decentraland/kernel/apis/user_identity.gen");
var UserIdentityServiceClient;
(function (UserIdentityServiceClient) {
    function create(clientPort) {
        return codegen.loadService(clientPort, user_identity_gen_1.UserIdentityServiceDefinition);
    }
    UserIdentityServiceClient.create = create;
    function createLegacy(clientPort) {
        const originalService = codegen.loadService(clientPort, user_identity_gen_1.UserIdentityServiceDefinition);
        return {
            ...originalService,
            async getUserPublicKey() {
                const realResponse = await originalService.getUserPublicKey({});
                return realResponse.address || null;
            },
            async getUserData() {
                const realResponse = await originalService.getUserData({});
                if (!realResponse.data) {
                    return null;
                }
                return {
                    ...realResponse.data,
                    avatar: {
                        ...realResponse.data.avatar,
                        snapshots: realResponse.data.avatar.snapshots
                    },
                    publicKey: realResponse.data.publicKey || null
                };
            }
        };
    }
    UserIdentityServiceClient.createLegacy = createLegacy;
})(UserIdentityServiceClient = exports.UserIdentityServiceClient || (exports.UserIdentityServiceClient = {}));
