// @dcl/scene-runtime/dist/worker/client/Web3Provider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLegacyWeb3Provider = void 0;
const codegen = __importStar(require("@dcl/rpc/dist/codegen"));
const ethereum_controller_gen_1 = require("@dcl/protocol/out-ts/decentraland/kernel/apis/ethereum_controller.gen");
function createLegacyWeb3Provider(clientPort) {
    const originalService = codegen.loadService(clientPort, ethereum_controller_gen_1.EthereumControllerServiceDefinition);
    async function request(message) {
        const response = await originalService.sendAsync({
            id: message.id,
            method: message.method,
            jsonParams: JSON.stringify(message.params)
        });
        return JSON.parse(response.jsonAnyResponse);
    }
    return {
        async getProvider() {
            return {
                // @internal
                send(message, callback) {
                    if (message && callback && callback instanceof Function) {
                        request(message)
                            .then((x) => callback(null, x))
                            .catch(callback);
                    }
                    else {
                        throw new Error('Decentraland provider only allows async calls');
                    }
                },
                sendAsync(message, callback) {
                    request(message)
                        .then((x) => callback(null, x))
                        .catch(callback);
                }
            };
        }
    };
}
exports.createLegacyWeb3Provider = createLegacyWeb3Provider;
