// @dcl/legacy-ecs/dist/ecs/ComponentGroup.js
import { getComponentName } from './Component';
/**
 * @public
 */
var ComponentGroup = /** @class */ (function () {
    function ComponentGroup() {
        var requires = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            requires[_i] = arguments[_i];
        }
        this.entities = [];
        this.active = false;
        this._requiresNames = [];
        // validate requires list
        if (!requires) {
            throw new Error('ComponentGroup: Could not load the requires list');
        }
        if (!(requires instanceof Array)) {
            throw new Error('ComponentGroup: requires list is not an Array');
        }
        Object.defineProperty(this, 'requires', {
            get: function () {
                return requires.slice();
            }
        });
        Object.defineProperty(this, 'requiresNames', {
            get: function () {
                return this._requiresNames.slice();
            }
        });
        var _loop_1 = function (ix) {
            var component = requires[ix];
            var name = null;
            if (!component) {
                throw new Error("ComponentGroup: the required component at location " + ix + " is invalid");
            }
            try {
                name = getComponentName(component);
            }
            catch (e) {
                throw new Error("ComponentGroup: the required component at location " + ix + " is not registered as a @Component. Remember to provide the class of the component, not the name");
            }
            if (this_1._requiresNames.some(function ($) { return $ === name; })) {
                throw new Error("ComponentGroup: the required component list has a repeated name " + name);
            }
            this_1._requiresNames.push(name);
        };
        var this_1 = this;
        for (var ix = 0; ix < requires.length; ix++) {
            _loop_1(ix);
        }
    }
    ComponentGroup.prototype.hasEntity = function (entity) {
        if (!entity.isAddedToEngine())
            return false;
        return this.entities.indexOf(entity) !== -1;
    };
    // @internal
    ComponentGroup.prototype.addEntity = function (entity) {
        if (!entity.isAddedToEngine()) {
            throw new TypeError('ComponentGroup: Cannot add a entity that is not added to the engine');
        }
        if (this.entities.indexOf(entity) === -1) {
            ;
            this.entities.push(entity);
        }
    };
    // @internal
    ComponentGroup.prototype.removeEntity = function (entity) {
        var id = this.entities.indexOf(entity);
        if (id !== -1) {
            ;
            this.entities.splice(id, 1);
        }
    };
    // @internal
    ComponentGroup.prototype.componentRemoved = function (entity, component) {
        if (this._requiresNames.indexOf(component) !== -1) {
            this.removeEntity(entity);
        }
    };
    // @internal
    ComponentGroup.prototype.meetsRequirements = function (entity) {
        for (var i = 0; i < this._requiresNames.length; i++) {
            var componentName = this._requiresNames[i];
            if (!(componentName in entity.components)) {
                return false;
            }
        }
        return true;
    };
    return ComponentGroup;
}());
export { ComponentGroup };
