// @dcl/legacy-ecs/dist/ecs/EventManager.js
import { error, newId } from './helpers';
var eventNameSymbol = '__event_name__';
var takenEventNames = [];
function getEventNameFromConstructor(ctor) {
    if (!(eventNameSymbol in ctor) || typeof ctor[eventNameSymbol] !== 'string') {
        throw new Error('The EventConstructor is not registered');
    }
    return ctor[eventNameSymbol];
}
/**
 * @public
 */
var EventManager = /** @class */ (function () {
    function EventManager() {
        this.listeners = {};
    }
    EventManager.prototype.addListener = function (eventClass, listener, listenerFunction) {
        if (!eventClass || typeof eventClass !== 'function') {
            throw new Error('Invalid EventConstructor');
        }
        var eventName = getEventNameFromConstructor(eventClass);
        var listeners = this.listeners[eventName];
        if (!listeners) {
            listeners = this.listeners[eventName] = [];
        }
        for (var i = 0; i < listeners.length; i++) {
            var x = listeners[i];
            if (x.listener === listener) {
                throw new Error('The provided listener is already registered');
            }
        }
        listeners.push({
            listener: listener,
            fn: listenerFunction
        });
        return this;
    };
    EventManager.prototype.removeListener = function (listener, eventClass) {
        if (!eventClass || typeof eventClass !== 'function') {
            throw new Error('Invalid EventConstructor');
        }
        var eventName = getEventNameFromConstructor(eventClass);
        var listeners = this.listeners[eventName];
        if (!listeners) {
            return false;
        }
        for (var i = 0; i < listeners.length; i++) {
            var x = listeners[i];
            if (x.listener === listener) {
                listeners.splice(i, 1);
                return true;
            }
        }
        return false;
    };
    EventManager.prototype.fireEvent = function (event) {
        var eventName = getEventNameFromConstructor(event.constructor);
        var listeners = this.listeners[eventName];
        if (listeners) {
            for (var i = 0; i < listeners.length; i++) {
                try {
                    var l = listeners[i];
                    l.fn.call(l.listener, event);
                }
                catch (e) {
                    // TODO: e may not be an instance of Error
                    error(e);
                }
            }
        }
        return this;
    };
    return EventManager;
}());
export { EventManager };
/**
 * @public
 */
export function EventConstructor() {
    var eventName = newId('EV');
    if (takenEventNames.indexOf(eventName) !== -1) {
        throw new Error("The event name " + eventName + " is already taken");
    }
    takenEventNames.push(eventName);
    // eslint-disable-next-line @typescript-eslint/ban-types
    return function (target) {
        ;
        target[eventNameSymbol] = eventName;
        return target;
    };
}
