// @dcl/legacy-ecs/dist/ecs/helpers.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var lastGeneratedId = 0;
/**
 * Log function. Only works in debug mode, otherwise it does nothing.
 * @param args - any loggable parameter
 * @public
 */
export function log() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    if (typeof dcl !== 'undefined') {
        dcl.log.apply(dcl, __spreadArray([], __read(args), false));
    }
    else {
        console.log.apply(console, __spreadArray(['DEBUG:'], __read(args), false));
    }
}
/**
 * Error function. Prints a console error. Only works in debug mode, otherwise it does nothing.
 * @param error - string or Error object.
 * @param data - any debug information.
 * @public
 */
export function error(error, data) {
    if (typeof dcl !== 'undefined') {
        dcl.error(error, data);
    }
    else {
        console.error('ERROR:', error, data);
    }
}
/**
 * Generates a new prefixed id
 * @public
 */
export function newId(type) {
    lastGeneratedId++;
    if (type.length === 0)
        throw new Error('newId(type: string): type cannot be empty');
    return type + lastGeneratedId.toString(36);
}
/**
 * @internal
 */
export function uuid() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        var r = (Math.random() * 16) | 0;
        var v = c === 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}
/**
 * Returns an array of the given size filled with element built from the given constructor and the paramters
 * @param size - the number of element to construct and put in the array
 * @param itemBuilder - a callback responsible for creating new instance of item. Called once per array entry.
 * @returns a new array filled with new objects
 * @internal
 */
export function buildArray(size, itemBuilder) {
    var a = [];
    for (var i = 0; i < size; ++i) {
        a.push(itemBuilder());
    }
    return a;
}
/**
 * @public
 */
export function openExternalURL(url) {
    if (typeof dcl !== 'undefined') {
        dcl.openExternalUrl(url);
    }
    else {
        error('ERROR: openExternalURL dcl is undefined');
    }
}
/**
 * Popup NFT info dialog
 * @param scr - 'ethereum://contractAddress/tokenID'
 * @param comment - optional. add a comment.
 * @public
 */
export function openNFTDialog(scr, comment) {
    if (comment === void 0) { comment = null; }
    if (typeof dcl !== 'undefined') {
        var regex = /ethereum:\/\/(.+)\/(.+)/;
        var matches = scr.match(regex);
        if (!matches || matches.length < 3) {
            return;
        }
        dcl.openNFTDialog(matches[1], matches[2], comment);
    }
    else {
        error('ERROR: openNFTDialog dcl is undefined');
    }
}
