// @dcl/legacy-ecs/dist/ecs/IEntity.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { EventConstructor } from './EventManager';
/**
 * @public
 */
var ComponentRemoved = /** @class */ (function () {
    function ComponentRemoved(entity, componentName, component) {
        this.entity = entity;
        this.componentName = componentName;
        this.component = component;
        // stub
    }
    ComponentRemoved = __decorate([
        EventConstructor(),
        __metadata("design:paramtypes", [Object, String, Object])
    ], ComponentRemoved);
    return ComponentRemoved;
}());
export { ComponentRemoved };
/**
 * @public
 */
var ComponentAdded = /** @class */ (function () {
    function ComponentAdded(entity, componentName, classId) {
        this.entity = entity;
        this.componentName = componentName;
        this.classId = classId;
        // stub
    }
    ComponentAdded = __decorate([
        EventConstructor(),
        __metadata("design:paramtypes", [Object, String, Object])
    ], ComponentAdded);
    return ComponentAdded;
}());
export { ComponentAdded };
/**
 * @public
 */
var ParentChanged = /** @class */ (function () {
    function ParentChanged(entity, parent) {
        this.entity = entity;
        this.parent = parent;
        // stub
    }
    ParentChanged = __decorate([
        EventConstructor(),
        __metadata("design:paramtypes", [Object, Object])
    ], ParentChanged);
    return ParentChanged;
}());
export { ParentChanged };
