// @dcl/legacy-ecs/dist/ecs/UIValue.js
/**
 * @public
 */
export var UIValueType;
(function (UIValueType) {
    UIValueType[UIValueType["PERCENT"] = 0] = "PERCENT";
    UIValueType[UIValueType["PIXELS"] = 1] = "PIXELS";
})(UIValueType || (UIValueType = {}));
/**
 * @public
 */
var UIValue = /** @class */ (function () {
    function UIValue(value) {
        this.type = UIValueType.PIXELS;
        if (typeof value === 'string') {
            var valueAsString = value;
            if (valueAsString.indexOf('px') > -1) {
                this.type = UIValueType.PIXELS;
            }
            else if (valueAsString.indexOf('%') > -1) {
                this.type = UIValueType.PERCENT;
            }
            this.value = parseFloat(valueAsString);
        }
        else {
            this.value = value;
        }
    }
    UIValue.prototype.toString = function () {
        var result = this.value.toString();
        if (this.type === UIValueType.PERCENT) {
            result += '%';
        }
        else {
            result += 'px';
        }
        return result;
    };
    return UIValue;
}());
export { UIValue };
