// dcl-social-client/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PresenceType = exports.CursorDirection = exports.MessageType = exports.MessageStatus = exports.ConversationType = exports.CHANNEL_TYPE = void 0;
exports.CHANNEL_TYPE = 'channel';
var ConversationType;
(function (ConversationType) {
    ConversationType[ConversationType["GROUP"] = 0] = "GROUP";
    ConversationType[ConversationType["DIRECT"] = 1] = "DIRECT";
    ConversationType[ConversationType["CHANNEL"] = 2] = "CHANNEL";
})(ConversationType = exports.ConversationType || (exports.ConversationType = {}));
var MessageStatus;
(function (MessageStatus) {
    MessageStatus["READ"] = "read";
    MessageStatus["UNREAD"] = "unread";
})(MessageStatus = exports.MessageStatus || (exports.MessageStatus = {}));
var MessageType;
(function (MessageType) {
    MessageType["TEXT"] = "m.text";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
var CursorDirection;
(function (CursorDirection) {
    CursorDirection[CursorDirection["BACKWARDS"] = 0] = "BACKWARDS";
    CursorDirection[CursorDirection["FORWARDS"] = 1] = "FORWARDS";
})(CursorDirection = exports.CursorDirection || (exports.CursorDirection = {}));
var PresenceType;
(function (PresenceType) {
    PresenceType["OFFLINE"] = "offline";
    PresenceType["ONLINE"] = "online";
    PresenceType["UNAVAILABLE"] = "unavailable"; // For example, idle
})(PresenceType = exports.PresenceType || (exports.PresenceType = {}));
