// @dcl/protocol/out-ts/google/api/http.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "google.api";

export interface Http {
  rules: HttpRule[];
}

export interface HttpRule {
  pattern?:
    | { $case: "get"; get: string }
    | { $case: "put"; put: string }
    | { $case: "post"; post: string }
    | { $case: "delete"; delete: string }
    | { $case: "patch"; patch: string }
    | { $case: "custom"; custom: CustomHttpPattern };
  selector: string;
  body: string;
  additionalBindings: HttpRule[];
}

export interface CustomHttpPattern {
  kind: string;
  path: string;
}

function createBaseHttp(): Http {
  return { rules: [] };
}

export const Http = {
  encode(message: Http, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    for (const v of message.rules) {
      HttpRule.encode(v!, writer.uint32(10).fork()).ldelim();
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): Http {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseHttp();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.rules.push(HttpRule.decode(reader, reader.uint32()));
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): Http {
    return { rules: Array.isArray(object?.rules) ? object.rules.map((e: any) => HttpRule.fromJSON(e)) : [] };
  },

  toJSON(message: Http): unknown {
    const obj: any = {};
    if (message.rules) {
      obj.rules = message.rules.map((e) => e ? HttpRule.toJSON(e) : undefined);
    } else {
      obj.rules = [];
    }
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<Http>, I>>(object: I): Http {
    const message = createBaseHttp();
    message.rules = object.rules?.map((e) => HttpRule.fromPartial(e)) || [];
    return message;
  },
};

function createBaseHttpRule(): HttpRule {
  return { pattern: undefined, selector: "", body: "", additionalBindings: [] };
}

export const HttpRule = {
  encode(message: HttpRule, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.pattern?.$case === "get") {
      writer.uint32(18).string(message.pattern.get);
    }
    if (message.pattern?.$case === "put") {
      writer.uint32(26).string(message.pattern.put);
    }
    if (message.pattern?.$case === "post") {
      writer.uint32(34).string(message.pattern.post);
    }
    if (message.pattern?.$case === "delete") {
      writer.uint32(42).string(message.pattern.delete);
    }
    if (message.pattern?.$case === "patch") {
      writer.uint32(50).string(message.pattern.patch);
    }
    if (message.pattern?.$case === "custom") {
      CustomHttpPattern.encode(message.pattern.custom, writer.uint32(66).fork()).ldelim();
    }
    if (message.selector !== "") {
      writer.uint32(10).string(message.selector);
    }
    if (message.body !== "") {
      writer.uint32(58).string(message.body);
    }
    for (const v of message.additionalBindings) {
      HttpRule.encode(v!, writer.uint32(90).fork()).ldelim();
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): HttpRule {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseHttpRule();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 2:
          message.pattern = { $case: "get", get: reader.string() };
          break;
        case 3:
          message.pattern = { $case: "put", put: reader.string() };
          break;
        case 4:
          message.pattern = { $case: "post", post: reader.string() };
          break;
        case 5:
          message.pattern = { $case: "delete", delete: reader.string() };
          break;
        case 6:
          message.pattern = { $case: "patch", patch: reader.string() };
          break;
        case 8:
          message.pattern = { $case: "custom", custom: CustomHttpPattern.decode(reader, reader.uint32()) };
          break;
        case 1:
          message.selector = reader.string();
          break;
        case 7:
          message.body = reader.string();
          break;
        case 11:
          message.additionalBindings.push(HttpRule.decode(reader, reader.uint32()));
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): HttpRule {
    return {
      pattern: isSet(object.get)
        ? { $case: "get", get: String(object.get) }
        : isSet(object.put)
        ? { $case: "put", put: String(object.put) }
        : isSet(object.post)
        ? { $case: "post", post: String(object.post) }
        : isSet(object.delete)
        ? { $case: "delete", delete: String(object.delete) }
        : isSet(object.patch)
        ? { $case: "patch", patch: String(object.patch) }
        : isSet(object.custom)
        ? { $case: "custom", custom: CustomHttpPattern.fromJSON(object.custom) }
        : undefined,
      selector: isSet(object.selector) ? String(object.selector) : "",
      body: isSet(object.body) ? String(object.body) : "",
      additionalBindings: Array.isArray(object?.additionalBindings)
        ? object.additionalBindings.map((e: any) => HttpRule.fromJSON(e))
        : [],
    };
  },

  toJSON(message: HttpRule): unknown {
    const obj: any = {};
    message.pattern?.$case === "get" && (obj.get = message.pattern?.get);
    message.pattern?.$case === "put" && (obj.put = message.pattern?.put);
    message.pattern?.$case === "post" && (obj.post = message.pattern?.post);
    message.pattern?.$case === "delete" && (obj.delete = message.pattern?.delete);
    message.pattern?.$case === "patch" && (obj.patch = message.pattern?.patch);
    message.pattern?.$case === "custom" &&
      (obj.custom = message.pattern?.custom ? CustomHttpPattern.toJSON(message.pattern?.custom) : undefined);
    message.selector !== undefined && (obj.selector = message.selector);
    message.body !== undefined && (obj.body = message.body);
    if (message.additionalBindings) {
      obj.additionalBindings = message.additionalBindings.map((e) => e ? HttpRule.toJSON(e) : undefined);
    } else {
      obj.additionalBindings = [];
    }
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<HttpRule>, I>>(object: I): HttpRule {
    const message = createBaseHttpRule();
    if (object.pattern?.$case === "get" && object.pattern?.get !== undefined && object.pattern?.get !== null) {
      message.pattern = { $case: "get", get: object.pattern.get };
    }
    if (object.pattern?.$case === "put" && object.pattern?.put !== undefined && object.pattern?.put !== null) {
      message.pattern = { $case: "put", put: object.pattern.put };
    }
    if (object.pattern?.$case === "post" && object.pattern?.post !== undefined && object.pattern?.post !== null) {
      message.pattern = { $case: "post", post: object.pattern.post };
    }
    if (object.pattern?.$case === "delete" && object.pattern?.delete !== undefined && object.pattern?.delete !== null) {
      message.pattern = { $case: "delete", delete: object.pattern.delete };
    }
    if (object.pattern?.$case === "patch" && object.pattern?.patch !== undefined && object.pattern?.patch !== null) {
      message.pattern = { $case: "patch", patch: object.pattern.patch };
    }
    if (object.pattern?.$case === "custom" && object.pattern?.custom !== undefined && object.pattern?.custom !== null) {
      message.pattern = { $case: "custom", custom: CustomHttpPattern.fromPartial(object.pattern.custom) };
    }
    message.selector = object.selector ?? "";
    message.body = object.body ?? "";
    message.additionalBindings = object.additionalBindings?.map((e) => HttpRule.fromPartial(e)) || [];
    return message;
  },
};

function createBaseCustomHttpPattern(): CustomHttpPattern {
  return { kind: "", path: "" };
}

export const CustomHttpPattern = {
  encode(message: CustomHttpPattern, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.kind !== "") {
      writer.uint32(10).string(message.kind);
    }
    if (message.path !== "") {
      writer.uint32(18).string(message.path);
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): CustomHttpPattern {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseCustomHttpPattern();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.kind = reader.string();
          break;
        case 2:
          message.path = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): CustomHttpPattern {
    return { kind: isSet(object.kind) ? String(object.kind) : "", path: isSet(object.path) ? String(object.path) : "" };
  },

  toJSON(message: CustomHttpPattern): unknown {
    const obj: any = {};
    message.kind !== undefined && (obj.kind = message.kind);
    message.path !== undefined && (obj.path = message.path);
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<CustomHttpPattern>, I>>(object: I): CustomHttpPattern {
    const message = createBaseCustomHttpPattern();
    message.kind = object.kind ?? "";
    message.path = object.path ?? "";
    return message;
  },
};

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
