// @dcl/legacy-ecs/dist/decentraland/AnimationState.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { ObservableComponent } from '../ecs/Component';
import { newId } from '../ecs/helpers';
var defaultParams = {
    looping: true,
    speed: 1.0,
    weight: 1.0,
    layer: 0
};
/**
 * @public
 */
var AnimationState = /** @class */ (function (_super) {
    __extends(AnimationState, _super);
    function AnimationState(clip, params) {
        if (params === void 0) { params = defaultParams; }
        var _this = _super.call(this) || this;
        // @internal
        _this.isAnimationClip = true;
        /**
         * Does the animation loop?, default: true
         */
        _this.looping = defaultParams.looping;
        /**
         * Weight of the animation, values from 0 to 1, used to blend several animations. default: 1
         */
        _this.weight = defaultParams.weight;
        /**
         * Is the animation playing? default: true
         */
        _this.playing = false;
        /**
         * Does any anyone asked to reset the animation? default: false
         */
        _this.shouldReset = false;
        /**
         * The animation speed
         */
        _this.speed = defaultParams.speed;
        // @internal
        _this.name = newId('AnimClip');
        /**
         * Layering allows you to have two or more levels of animation on an object's parameters at the same time
         */
        _this.layer = defaultParams.layer;
        _this.clip = clip;
        _this.setParams(__assign({}, params));
        return _this;
    }
    /**
     * Sets the clip parameters
     */
    AnimationState.prototype.setParams = function (params) {
        this.looping = params.looping !== undefined ? params.looping : this.looping;
        this.speed = params.speed || this.speed;
        this.weight = params.weight || this.weight;
        this.layer = params.layer || this.layer;
        return this;
    };
    AnimationState.prototype.toJSON = function () {
        var ret = JSON.parse(JSON.stringify(_super.prototype.toJSON.call(this)));
        if (this.shouldReset) {
            this.shouldReset = false;
        }
        return ret;
    };
    /**
     * Starts the animation
     */
    AnimationState.prototype.play = function (reset) {
        var _a;
        if (reset === void 0) { reset = false; }
        (_a = this.owner) === null || _a === void 0 ? void 0 : _a.play(this, reset);
    };
    /**
     * Pauses the animation
     */
    AnimationState.prototype.pause = function () {
        var _a;
        (_a = this.owner) === null || _a === void 0 ? void 0 : _a.pause(this);
    };
    /**
     * Resets the animation state to the frame 0
     */
    AnimationState.prototype.reset = function () {
        this.shouldReset = true;
    };
    /**
     * Resets and pauses the animation
     */
    AnimationState.prototype.stop = function () {
        var _a;
        (_a = this.owner) === null || _a === void 0 ? void 0 : _a.stop(this);
    };
    __decorate([
        ObservableComponent.readonly,
        __metadata("design:type", String)
    ], AnimationState.prototype, "clip", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], AnimationState.prototype, "looping", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], AnimationState.prototype, "weight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], AnimationState.prototype, "playing", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], AnimationState.prototype, "shouldReset", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], AnimationState.prototype, "speed", void 0);
    __decorate([
        ObservableComponent.readonly,
        __metadata("design:type", String)
    ], AnimationState.prototype, "name", void 0);
    return AnimationState;
}(ObservableComponent));
export { AnimationState };
