// @dcl/legacy-ecs/dist/decentraland/Audio.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { DisposableComponent, ObservableComponent, Component, getComponentId } from '../ecs/Component';
import { CLASS_ID } from './Components';
/**
 * @public
 */
var AudioClip = /** @class */ (function (_super) {
    __extends(AudioClip, _super);
    function AudioClip(url) {
        var _this = _super.call(this) || this;
        /**
         * Is this clip looping by default?
         */
        _this.loop = false;
        /**
         * Clip's master volume. This volume affects all the AudioSources.
         * Valid ranges from 0 to 1
         */
        _this.volume = 1;
        _this.url = url;
        return _this;
    }
    __decorate([
        ObservableComponent.readonly,
        __metadata("design:type", String)
    ], AudioClip.prototype, "url", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], AudioClip.prototype, "loop", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], AudioClip.prototype, "loadingCompleteEventId", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], AudioClip.prototype, "volume", void 0);
    AudioClip = __decorate([
        DisposableComponent('engine.AudioClip', CLASS_ID.AUDIO_CLIP),
        __metadata("design:paramtypes", [String])
    ], AudioClip);
    return AudioClip;
}(ObservableComponent));
export { AudioClip };
/**
 * @public
 */
var AudioSource = /** @class */ (function (_super) {
    __extends(AudioSource, _super);
    function AudioSource(audioClip) {
        var _this = _super.call(this) || this;
        _this.audioClip = audioClip;
        /**
         * Is this clip looping by default?
         */
        _this.loop = false;
        /**
         * Clip's master volume. This volume affects all the AudioSources.
         * Valid ranges from 0 to 1
         */
        _this.volume = 1;
        /**
         * Is this AudioSource playing?
         */
        _this.playing = false;
        /**
         * Pitch, default: 1.0, range from 0.0 to MaxFloat
         */
        _this.pitch = 1.0;
        /**
         * Timestamp of when the playOnce was executed to see if we need to start over the audio
         */
        _this.playedAtTimestamp = Date.now();
        if (!(audioClip instanceof AudioClip)) {
            throw new Error("Trying to create AudioSource(AudioClip) with an invalid AudioClip");
        }
        _this.audioClipId = getComponentId(audioClip);
        return _this;
    }
    /**
     * Disables the looping and plays the current source once.
     * If the sound was playing, it stops and starts over.
     */
    AudioSource.prototype.playOnce = function () {
        this.playing = true;
        this.dirty = true;
        this.playedAtTimestamp = Date.now();
        this.data.nonce = Math.random();
        return this;
    };
    __decorate([
        ObservableComponent.readonly,
        __metadata("design:type", String)
    ], AudioSource.prototype, "audioClipId", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], AudioSource.prototype, "loop", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], AudioSource.prototype, "volume", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], AudioSource.prototype, "playing", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], AudioSource.prototype, "pitch", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], AudioSource.prototype, "playedAtTimestamp", void 0);
    AudioSource = __decorate([
        Component('engine.AudioSource', CLASS_ID.AUDIO_SOURCE),
        __metadata("design:paramtypes", [AudioClip])
    ], AudioSource);
    return AudioSource;
}(ObservableComponent));
export { AudioSource };
/**
 * @public
 */
var AudioStream = /** @class */ (function (_super) {
    __extends(AudioStream, _super);
    function AudioStream(url) {
        var _this = _super.call(this) || this;
        _this.playing = false;
        _this.volume = 1;
        _this.url = url;
        _this.playing = true;
        return _this;
    }
    __decorate([
        ObservableComponent.readonly,
        __metadata("design:type", String)
    ], AudioStream.prototype, "url", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], AudioStream.prototype, "playing", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], AudioStream.prototype, "volume", void 0);
    AudioStream = __decorate([
        Component('engine.AudioStream', CLASS_ID.AUDIO_STREAM),
        __metadata("design:paramtypes", [String])
    ], AudioStream);
    return AudioStream;
}(ObservableComponent));
export { AudioStream };
