// @dcl/legacy-ecs/dist/decentraland/Events.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { EventConstructor } from '../ecs/EventManager';
import { Observable } from '../ecs/Observable';
import { DisposableComponent } from '../ecs/Component';
/**
 * @public
 */
var UUIDEvent = /** @class */ (function () {
    function UUIDEvent(uuid, payload) {
        this.uuid = uuid;
        this.payload = payload;
    }
    UUIDEvent = __decorate([
        EventConstructor(),
        __metadata("design:paramtypes", [String, Object])
    ], UUIDEvent);
    return UUIDEvent;
}());
export { UUIDEvent };
/**
 * @public
 */
var RaycastResponse = /** @class */ (function () {
    function RaycastResponse(payload) {
        this.payload = payload;
    }
    RaycastResponse = __decorate([
        EventConstructor(),
        __metadata("design:paramtypes", [Object])
    ], RaycastResponse);
    return RaycastResponse;
}());
export { RaycastResponse };
/**
 * @public
 */
var PointerEvent = /** @class */ (function () {
    function PointerEvent(payload) {
        this.payload = payload;
    }
    PointerEvent = __decorate([
        EventConstructor(),
        __metadata("design:paramtypes", [Object])
    ], PointerEvent);
    return PointerEvent;
}());
export { PointerEvent };
var internalDcl;
/**
 * @internal
 * This function generates a callback that is passed to the Observable
 * constructor to subscribe to the events of the DecentralandInterface
 */
function createSubscriber(eventName) {
    return function () {
        if (internalDcl) {
            internalDcl.subscribe(eventName);
        }
    };
}
/**
 * This event is triggered when you change your camera between 1st and 3rd person
 * @public
 */
export var onCameraModeChangedObservable = new Observable(createSubscriber('cameraModeChanged'));
/**
 * This event is triggered when you change your camera between 1st and 3rd person
 * @public
 */
export var onIdleStateChangedObservable = new Observable(createSubscriber('idleStateChanged'));
/**
 * These events are triggered after your character enters the scene.
 * @public
 */
export var onEnterSceneObservable = new Observable(createSubscriber('onEnterScene'));
/** @public @deprecated Use onEnterSceneObservable instead. */
export var onEnterScene = onEnterSceneObservable;
/**
 * These events are triggered after your character leaves the scene.
 * @public
 */
export var onLeaveSceneObservable = new Observable(createSubscriber('onLeaveScene'));
/** @public @deprecated Use onLeaveSceneObservable instead. */
export var onLeaveScene = onLeaveSceneObservable;
/**
 * This event is triggered after all the resources of the scene were loaded (models, textures, etc...)
 * @public
 */
export var onSceneReadyObservable = new Observable(createSubscriber('sceneStart'));
/**
 * @public
 */
export var onPlayerExpressionObservable = new Observable(createSubscriber('playerExpression'));
/**
 * @public
 */
export var onPointerLockedStateChange = new Observable(createSubscriber('onPointerLock'));
/**
 * @public
 */
export var onVideoEvent = new Observable(createSubscriber('videoEvent'));
/**
 * @public
 */
export var onProfileChanged = new Observable(createSubscriber('profileChanged'));
/**
 * @public
 */
export var onPlayerConnectedObservable = new Observable(createSubscriber('playerConnected'));
/**
 * @public
 */
export var onPlayerDisconnectedObservable = new Observable(createSubscriber('playerDisconnected'));
/**
 * @public
 */
export var onRealmChangedObservable = new Observable(createSubscriber('onRealmChanged'));
/**
 * @public
 */
export var onPlayerClickedObservable = new Observable(createSubscriber('playerClicked'));
/**
 * @internal
 * This function adds _one_ listener to the onEvent event of dcl interface.
 * Leveraging a switch to route events to the Observable handlers.
 */
export function _initEventObservables(dcl) {
    // store internal reference to dcl, it is going to be used to subscribe to the events
    internalDcl = dcl;
    if (internalDcl) {
        internalDcl.onEvent(function (event) {
            switch (event.type) {
                case 'onEnterScene': {
                    onEnterSceneObservable.notifyObservers(event.data);
                    return;
                }
                case 'onLeaveScene': {
                    onLeaveSceneObservable.notifyObservers(event.data);
                    return;
                }
                case 'cameraModeChanged': {
                    onCameraModeChangedObservable.notifyObservers(event.data);
                    return;
                }
                case 'idleStateChanged': {
                    onIdleStateChangedObservable.notifyObservers(event.data);
                    return;
                }
                case 'sceneStart': {
                    onSceneReadyObservable.notifyObservers(event.data);
                    return;
                }
                case 'playerExpression': {
                    onPlayerExpressionObservable.notifyObservers(event.data);
                    return;
                }
                case 'videoEvent': {
                    var videoData = event.data;
                    var component = DisposableComponent.engine.disposableComponents[videoData.componentId];
                    if (component) {
                        component.update(videoData);
                    }
                    onVideoEvent.notifyObservers(videoData);
                    return;
                }
                case 'profileChanged': {
                    onProfileChanged.notifyObservers(event.data);
                    return;
                }
                case 'onPointerLock': {
                    onPointerLockedStateChange.notifyObservers(event.data);
                    return;
                }
                case 'playerConnected': {
                    onPlayerConnectedObservable.notifyObservers(event.data);
                    return;
                }
                case 'playerDisconnected': {
                    onPlayerDisconnectedObservable.notifyObservers(event.data);
                    return;
                }
                case 'onRealmChanged': {
                    onRealmChangedObservable.notifyObservers(event.data);
                    return;
                }
                case 'playerClicked': {
                    onPlayerClickedObservable.notifyObservers(event.data);
                    return;
                }
            }
        });
    }
}
