// @dcl/legacy-ecs/dist/decentraland/Gizmos.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, ObservableComponent } from '../ecs/Component';
import { CLASS_ID, OnUUIDEvent } from './Components';
/**
 * Gizmo identifiers
 * @public
 */
export var Gizmo;
(function (Gizmo) {
    Gizmo["MOVE"] = "MOVE";
    Gizmo["ROTATE"] = "ROTATE";
    Gizmo["SCALE"] = "SCALE";
    Gizmo["NONE"] = "NONE";
})(Gizmo || (Gizmo = {}));
/**
 * This event is triggered after the user finalizes dragging a gizmo.
 * @public
 */
var OnGizmoEvent = /** @class */ (function (_super) {
    __extends(OnGizmoEvent, _super);
    function OnGizmoEvent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'gizmoEvent';
        return _this;
    }
    __decorate([
        ObservableComponent.readonly,
        __metadata("design:type", String)
    ], OnGizmoEvent.prototype, "type", void 0);
    OnGizmoEvent = __decorate([
        Component('engine.gizmoEvent', CLASS_ID.UUID_CALLBACK)
    ], OnGizmoEvent);
    return OnGizmoEvent;
}(OnUUIDEvent));
export { OnGizmoEvent };
/**
 * Enables gizmos in the entity. Gizmos only work in EDITOR, PREVIEW or DEBUG modes.
 * @public
 */
var Gizmos = /** @class */ (function (_super) {
    __extends(Gizmos, _super);
    function Gizmos() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Enable position gizmo
         */
        _this.position = true;
        /**
         * Enable rotation gizmo
         */
        _this.rotation = true;
        /**
         * Enable scale gizmo
         */
        _this.scale = true;
        /**
         * Cycle through gizmos using click.
         */
        _this.cycle = true;
        /**
         * Align the gizmos to match the local reference system
         */
        _this.localReference = false;
        return _this;
    }
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], Gizmos.prototype, "position", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], Gizmos.prototype, "rotation", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], Gizmos.prototype, "scale", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], Gizmos.prototype, "cycle", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], Gizmos.prototype, "selectedGizmo", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], Gizmos.prototype, "localReference", void 0);
    Gizmos = __decorate([
        Component('engine.gizmos', CLASS_ID.GIZMOS)
    ], Gizmos);
    return Gizmos;
}(ObservableComponent));
export { Gizmos };
