// @dcl/legacy-ecs/dist/decentraland/Input.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Vector3 } from '@dcl/ecs-math';
import { InputEventType } from './Types';
import { Component, DisposableComponent } from '../ecs/Component';
/**
 * @public
 */
export var ActionButton;
(function (ActionButton) {
    ActionButton["POINTER"] = "POINTER";
    ActionButton["PRIMARY"] = "PRIMARY";
    ActionButton["SECONDARY"] = "SECONDARY";
    ActionButton["ANY"] = "ANY";
    ActionButton["FORWARD"] = "FORWARD";
    ActionButton["BACKWARD"] = "BACKWARD";
    ActionButton["RIGHT"] = "RIGHT";
    ActionButton["LEFT"] = "LEFT";
    ActionButton["JUMP"] = "JUMP";
    ActionButton["WALK"] = "WALK";
    ActionButton["ACTION_3"] = "ACTION_3";
    ActionButton["ACTION_4"] = "ACTION_4";
    ActionButton["ACTION_5"] = "ACTION_5";
    ActionButton["ACTION_6"] = "ACTION_6";
})(ActionButton || (ActionButton = {}));
/**
 * @public
 */
var PointerEventComponent = /** @class */ (function () {
    function PointerEventComponent(callback) {
        this.callback = callback;
        if (!callback || !('apply' in callback) || !('call' in callback)) {
            throw new Error('Callback is not a function');
        }
        Input.ensureInstance();
    }
    return PointerEventComponent;
}());
export { PointerEventComponent };
/**
 * @public
 */
var GlobalPointerDown = /** @class */ (function (_super) {
    __extends(GlobalPointerDown, _super);
    function GlobalPointerDown() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GlobalPointerDown = __decorate([
        Component('pointerDown')
    ], GlobalPointerDown);
    return GlobalPointerDown;
}(PointerEventComponent));
export { GlobalPointerDown };
/**
 * @public
 */
var GlobalPointerUp = /** @class */ (function (_super) {
    __extends(GlobalPointerUp, _super);
    function GlobalPointerUp() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GlobalPointerUp = __decorate([
        Component('pointerUp')
    ], GlobalPointerUp);
    return GlobalPointerUp;
}(PointerEventComponent));
export { GlobalPointerUp };
/**
 * @public
 */
var Subscription = /** @class */ (function () {
    function Subscription(fn, useRaycast) {
        this.fn = fn;
        this.useRaycast = useRaycast;
    }
    return Subscription;
}());
export { Subscription };
/**
 * @public
 */
var Input = /** @class */ (function () {
    function Input() {
        // @internal
        this.buttonIdMapping = [
            ActionButton.POINTER,
            ActionButton.PRIMARY,
            ActionButton.SECONDARY,
            ActionButton.ANY,
            ActionButton.FORWARD,
            ActionButton.BACKWARD,
            ActionButton.RIGHT,
            ActionButton.LEFT,
            ActionButton.JUMP,
            ActionButton.WALK,
            ActionButton.ACTION_3,
            ActionButton.ACTION_4,
            ActionButton.ACTION_5,
            ActionButton.ACTION_6
        ];
        // @internal
        this.subscriptions = this.buttonIdMapping.reduce(function (acc, k) {
            var _a;
            return (__assign(__assign({}, acc), (_a = {}, _a[k] = { BUTTON_DOWN: [], BUTTON_UP: [] }, _a)));
        }, {});
        this.internalState = this.buttonIdMapping.reduce(function (acc, k) {
            var _a;
            return (__assign(__assign({}, acc), (_a = {}, _a[k] = { BUTTON_DOWN: false }, _a)));
        }, {});
    }
    Object.defineProperty(Input, "instance", {
        get: function () {
            Input.ensureInstance();
            return Input._instance;
        },
        enumerable: false,
        configurable: true
    });
    Input.ensureInstance = function () {
        if (!Input._instance) {
            Input._instance = new Input();
        }
    };
    /**
     * Allows to know if a button is pressed
     *
     * Returns true if the button is pressed
     * @param buttonId - The id of the button.
     */
    Input.prototype.isButtonPressed = function (buttonId) {
        return this.internalState[buttonId];
    };
    /**
     * Subscribes to an input event and triggers the provided callback.
     *
     * Returns a function that can be called to remove the subscription.
     * @param eventName - The name of the event (see InputEventKind).
     * @param buttonId - The id of the button.
     * @param useRaycast - Enables getting raycast information.
     * @param fn - A callback function to be called when the event is triggered.
     */
    Input.prototype.subscribe = function (eventName, buttonId, useRaycast, fn) {
        var _this = this;
        this.subscriptions[buttonId][eventName].push(new Subscription(fn, useRaycast));
        return function () {
            _this.unsubscribe(eventName, buttonId, fn);
        };
    };
    /**
     * Removes an existing input event subscription.
     * @param eventName - The name of the event (see InputEventKind).
     * @param buttonId - The id of the button.
     * @param fn - The callback function used when subscribing to the event.
     */
    Input.prototype.unsubscribe = function (eventName, buttonId, fn) {
        var index = this.getSubscriptionId(eventName, buttonId, fn);
        if (index > -1) {
            return this.subscriptions[buttonId][eventName].splice(index, 1);
        }
        return false;
    };
    Input.prototype.handlePointerEvent = function (data) {
        var button = this.getPointerById(data.buttonId);
        if (!button) {
            return;
        }
        var eventResult = __assign(__assign({}, data), { button: button, direction: new Vector3().copyFrom(data.direction), origin: new Vector3().copyFrom(data.origin), hit: undefined });
        var hit = data.hit
            ? __assign(__assign({}, data.hit), { hitPoint: new Vector3().copyFrom(data.hit.hitPoint), normal: new Vector3().copyFrom(data.hit.normal), worldNormal: new Vector3().copyFrom(data.hit.worldNormal) }) : undefined;
        if (data.type === InputEventType.DOWN) {
            this.internalState[button].BUTTON_DOWN = true;
            for (var i = 0; i < this.subscriptions[button]['BUTTON_DOWN'].length; i++) {
                var subscription = this.subscriptions[button]['BUTTON_DOWN'][i];
                // remove hit information when raycast is disabled
                if (subscription.useRaycast) {
                    eventResult.hit = hit;
                }
                else {
                    eventResult.hit = undefined;
                }
                subscription.fn(eventResult);
            }
            if (hit && hit.entityId && DisposableComponent.engine) {
                var entity = DisposableComponent.engine.entities[hit.entityId];
                var handler = entity && entity.getComponentOrNull(GlobalPointerDown);
                if (handler) {
                    eventResult.hit = hit;
                    handler.callback(eventResult);
                }
            }
        }
        else {
            this.internalState[button].BUTTON_DOWN = false;
            for (var i = 0; i < this.subscriptions[button]['BUTTON_UP'].length; i++) {
                var subscription = this.subscriptions[button]['BUTTON_UP'][i];
                // remove hit information when raycast is disabled
                if (subscription.useRaycast) {
                    eventResult.hit = hit;
                }
                else {
                    eventResult.hit = undefined;
                }
                subscription.fn(eventResult);
            }
            if (hit && hit.entityId && DisposableComponent.engine) {
                var entity = DisposableComponent.engine.entities[hit.entityId];
                var handler = entity && entity.getComponentOrNull(GlobalPointerUp);
                if (handler) {
                    eventResult.hit = hit;
                    handler.callback(eventResult);
                }
            }
        }
    };
    Input.prototype.getSubscriptionId = function (eventName, buttonId, fn) {
        for (var i = 0; i < this.subscriptions[buttonId][eventName].length; i++) {
            if (this.subscriptions[buttonId][eventName][i].fn === fn) {
                return i;
            }
        }
        return -1;
    };
    Input.prototype.getPointerById = function (id) {
        if (id < 0 || id >= this.buttonIdMapping.length) {
            return null;
        }
        var actionButton = this.buttonIdMapping[id];
        if (actionButton === ActionButton.ANY) {
            return null;
        }
        return actionButton;
    };
    return Input;
}());
export { Input };
