// @dcl/legacy-ecs/dist/decentraland/MessageBus.js
import { Observable } from '../ecs/Observable';
import { error } from '../ecs/helpers';
var communicationsController = null;
var communicationsControllerPromise = null;
var _messageObserver = null;
/**
 * @internal
 */
export function getMessageObserver() {
    if (!_messageObserver) {
        _messageObserver = new Observable();
    }
    return _messageObserver;
}
function ensureCommunicationsController() {
    if (!communicationsControllerPromise) {
        communicationsControllerPromise = dcl.loadModule('@decentraland/CommunicationsController', {});
        void communicationsControllerPromise.then(function ($) {
            communicationsController = $;
        });
        var observer_1 = getMessageObserver();
        dcl.subscribe('comms');
        dcl.onEvent(function (event) {
            if (event.type === 'comms') {
                observer_1.notifyObservers(event.data);
            }
        });
    }
    return communicationsControllerPromise;
}
/**
 * @public
 */
var MessageBus = /** @class */ (function () {
    function MessageBus() {
        var _this = this;
        this.messageQueue = [];
        this.connected = false;
        this.flushing = false;
        void ensureCommunicationsController().then(function () {
            _this.connected = true;
            _this.flush();
        });
    }
    MessageBus.prototype.on = function (message, callback) {
        return getMessageObserver().add(function (e) {
            try {
                var m = JSON.parse(e.message);
                if (m.message === message) {
                    callback(m.payload, e.sender);
                }
            }
            catch (e) {
                dcl.error('Error parsing comms message ' + (e.message || ''), e);
            }
        });
    };
    // @internal
    MessageBus.prototype.sendRaw = function (message) {
        this.messageQueue.push(message);
        if (this.connected) {
            this.flush();
        }
    };
    MessageBus.prototype.emit = function (message, payload) {
        var messageToSend = JSON.stringify({ message: message, payload: payload });
        this.sendRaw(messageToSend);
        getMessageObserver().notifyObservers({
            message: messageToSend,
            sender: 'self'
        });
    };
    MessageBus.prototype.flush = function () {
        var _this = this;
        if (this.messageQueue.length === 0)
            return;
        if (!this.connected)
            return;
        if (!communicationsController)
            return;
        if (this.flushing)
            return;
        var message = this.messageQueue.shift();
        this.flushing = true;
        dcl.callRpc(communicationsController.rpcHandle, 'send', [message]).then(function (_) {
            _this.flushing = false;
            _this.flush();
        }, function (e) {
            _this.flushing = false;
            error('Error flushing MessageBus', e);
        });
    };
    return MessageBus;
}());
export { MessageBus };
