// @dcl/legacy-ecs/dist/decentraland/PhysicsCast.js
import { Vector3, Matrix } from '@dcl/ecs-math';
import { uuid, log } from '../ecs/helpers';
import { Camera } from './Camera';
/**
 * @internal
 */
var QueryPrefix;
(function (QueryPrefix) {
    QueryPrefix["HitFirst"] = "rqhf";
    QueryPrefix["HitAll"] = "rqha";
})(QueryPrefix || (QueryPrefix = {}));
/**
 * @public
 */
var PhysicsCast = /** @class */ (function () {
    function PhysicsCast() {
        this.queries = {};
    }
    Object.defineProperty(PhysicsCast, "instance", {
        get: function () {
            PhysicsCast.ensureInstance();
            return PhysicsCast._instance;
        },
        enumerable: false,
        configurable: true
    });
    PhysicsCast.ensureInstance = function () {
        if (!PhysicsCast._instance) {
            PhysicsCast._instance = new PhysicsCast();
        }
    };
    PhysicsCast.prototype.getRayFromCamera = function (distance) {
        var rotation = Camera.instance.rotation;
        var rotationMat = Matrix.Identity();
        rotation.toRotationMatrix(rotationMat);
        var direction = Vector3.TransformCoordinates(Vector3.Forward(), rotationMat);
        var ray = {
            origin: Camera.instance.position,
            direction: direction,
            distance: distance
        };
        return ray;
    };
    PhysicsCast.prototype.getRayFromPositions = function (from, to) {
        var direction = to.subtract(from);
        var length = direction.length();
        var ray = {
            origin: from,
            direction: direction.normalize(),
            distance: length
        };
        return ray;
    };
    PhysicsCast.prototype.hitFirst = function (ray, hitCallback, id) {
        var queryId = typeof id === 'number' ? QueryPrefix.HitFirst + id : uuid();
        this.queries[queryId] = hitCallback;
        if (typeof dcl !== 'undefined') {
            dcl.query('raycast', { queryId: queryId, queryType: 'HitFirst', ray: ray });
        }
    };
    PhysicsCast.prototype.hitAll = function (ray, hitCallback, id) {
        var queryId = typeof id === 'number' ? QueryPrefix.HitAll + id : uuid();
        this.queries[queryId] = hitCallback;
        if (typeof dcl !== 'undefined') {
            dcl.query('raycast', { queryId: queryId, queryType: 'HitAll', ray: ray });
        }
    };
    PhysicsCast.prototype.hitFirstAvatar = function (_ray, _hitCallback) {
        log('not implemented yet');
    };
    PhysicsCast.prototype.hitAllAvatars = function (_ray, _hitCallback) {
        log('not implemented yet');
    };
    PhysicsCast.prototype.handleRaycastHitFirstResponse = function (response) {
        this.queries[response.payload.queryId](response.payload.payload);
        delete this.queries[response.payload.queryId];
    };
    PhysicsCast.prototype.handleRaycastHitAllResponse = function (response) {
        this.queries[response.payload.queryId](response.payload.payload);
        delete this.queries[response.payload.queryId];
    };
    return PhysicsCast;
}());
export { PhysicsCast };
