// @dcl/legacy-ecs/dist/decentraland/UIShapes.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Color4 } from '@dcl/ecs-math';
import { ObservableComponent, DisposableComponent, getComponentId } from '../ecs/Component';
import { CLASS_ID, OnUUIDEvent, Font } from './Components';
import { OnChanged } from './UIEvents';
/**
 * @public
 */
var UIShape = /** @class */ (function (_super) {
    __extends(UIShape, _super);
    function UIShape(parent) {
        var _this = _super.call(this) || this;
        /**
         * Defines if the entity and its children should be rendered
         */
        _this.name = null;
        _this.visible = true;
        _this.opacity = 1;
        _this.hAlign = 'center';
        _this.vAlign = 'center';
        _this.width = '100px';
        _this.height = '50px';
        _this.positionX = '0px';
        _this.positionY = '0px';
        _this.isPointerBlocker = true;
        if (parent) {
            _this._parent = parent;
            _this.data.parentComponent = getComponentId(parent);
        }
        return _this;
    }
    Object.defineProperty(UIShape.prototype, "parent", {
        get: function () {
            return this._parent;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(UIShape.prototype, "parentComponent", {
        // @internal
        get: function () {
            return this.data.parentComponent;
        },
        enumerable: false,
        configurable: true
    });
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Object)
    ], UIShape.prototype, "name", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIShape.prototype, "visible", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIShape.prototype, "opacity", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIShape.prototype, "hAlign", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIShape.prototype, "vAlign", void 0);
    __decorate([
        ObservableComponent.uiValue,
        __metadata("design:type", Object)
    ], UIShape.prototype, "width", void 0);
    __decorate([
        ObservableComponent.uiValue,
        __metadata("design:type", Object)
    ], UIShape.prototype, "height", void 0);
    __decorate([
        ObservableComponent.uiValue,
        __metadata("design:type", Object)
    ], UIShape.prototype, "positionX", void 0);
    __decorate([
        ObservableComponent.uiValue,
        __metadata("design:type", Object)
    ], UIShape.prototype, "positionY", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIShape.prototype, "isPointerBlocker", void 0);
    return UIShape;
}(ObservableComponent));
export { UIShape };
/**
 * @internal
 * NOTE(Brian): this should be deprecated
 */
var UIFullScreen = /** @class */ (function (_super) {
    __extends(UIFullScreen, _super);
    function UIFullScreen() {
        return _super.call(this, null) || this;
    }
    UIFullScreen = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_FULLSCREEN_SHAPE),
        __metadata("design:paramtypes", [])
    ], UIFullScreen);
    return UIFullScreen;
}(UIShape));
export { UIFullScreen };
/**
 * @internal
 * NOTE(Brian): this should be deprecated
 */
var UIWorldSpace = /** @class */ (function (_super) {
    __extends(UIWorldSpace, _super);
    function UIWorldSpace() {
        return _super.call(this, null) || this;
    }
    UIWorldSpace = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_WORLD_SPACE_SHAPE),
        __metadata("design:paramtypes", [])
    ], UIWorldSpace);
    return UIWorldSpace;
}(UIShape));
export { UIWorldSpace };
/**
 * @public
 */
var UICanvas = /** @class */ (function (_super) {
    __extends(UICanvas, _super);
    function UICanvas() {
        return _super.call(this, null) || this;
    }
    UICanvas = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_SCREEN_SPACE_SHAPE),
        __metadata("design:paramtypes", [])
    ], UICanvas);
    return UICanvas;
}(UIShape));
export { UICanvas };
/**
 * @public
 */
var UIContainerRect = /** @class */ (function (_super) {
    __extends(UIContainerRect, _super);
    function UIContainerRect() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.thickness = 0;
        _this.color = Color4.Clear();
        _this.alignmentUsesSize = true;
        return _this;
    }
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIContainerRect.prototype, "thickness", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIContainerRect.prototype, "color", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIContainerRect.prototype, "alignmentUsesSize", void 0);
    UIContainerRect = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_CONTAINER_RECT)
    ], UIContainerRect);
    return UIContainerRect;
}(UIShape));
export { UIContainerRect };
/**
 * @public
 */
export var UIStackOrientation;
(function (UIStackOrientation) {
    UIStackOrientation[UIStackOrientation["VERTICAL"] = 0] = "VERTICAL";
    UIStackOrientation[UIStackOrientation["HORIZONTAL"] = 1] = "HORIZONTAL";
})(UIStackOrientation || (UIStackOrientation = {}));
/**
 * @public
 */
var UIContainerStack = /** @class */ (function (_super) {
    __extends(UIContainerStack, _super);
    function UIContainerStack() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.adaptWidth = true;
        _this.adaptHeight = true;
        _this.color = Color4.Clear();
        _this.stackOrientation = UIStackOrientation.VERTICAL;
        _this.spacing = 0;
        return _this;
    }
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIContainerStack.prototype, "adaptWidth", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIContainerStack.prototype, "adaptHeight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIContainerStack.prototype, "color", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIContainerStack.prototype, "stackOrientation", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIContainerStack.prototype, "spacing", void 0);
    UIContainerStack = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_CONTAINER_STACK)
    ], UIContainerStack);
    return UIContainerStack;
}(UIShape));
export { UIContainerStack };
/**
 * @public
 */
var UIButton = /** @class */ (function (_super) {
    __extends(UIButton, _super);
    function UIButton() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.fontSize = 10;
        _this.fontWeight = 'normal';
        _this.thickness = 0;
        _this.cornerRadius = 0;
        _this.color = Color4.White();
        _this.background = Color4.White();
        _this.paddingTop = 0;
        _this.paddingRight = 0;
        _this.paddingBottom = 0;
        _this.paddingLeft = 0;
        _this.shadowBlur = 0;
        _this.shadowOffsetX = 0;
        _this.shadowOffsetY = 0;
        _this.shadowColor = Color4.Black();
        _this.text = 'button';
        return _this;
    }
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "fontSize", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIButton.prototype, "fontWeight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "thickness", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "cornerRadius", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIButton.prototype, "color", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIButton.prototype, "background", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "paddingTop", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "paddingRight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "paddingBottom", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "paddingLeft", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "shadowBlur", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "shadowOffsetX", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIButton.prototype, "shadowOffsetY", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIButton.prototype, "shadowColor", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIButton.prototype, "text", void 0);
    UIButton = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_BUTTON_SHAPE)
    ], UIButton);
    return UIButton;
}(UIShape));
export { UIButton };
/**
 * @public
 */
var UIText = /** @class */ (function (_super) {
    __extends(UIText, _super);
    function UIText() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.outlineWidth = 0;
        _this.outlineColor = Color4.White();
        _this.color = Color4.White();
        _this.fontSize = 10;
        _this.fontAutoSize = false;
        _this.value = '';
        _this.lineSpacing = 0;
        _this.lineCount = 0;
        _this.adaptWidth = false;
        _this.adaptHeight = false;
        _this.textWrapping = false;
        _this.shadowBlur = 0;
        _this.shadowOffsetX = 0;
        _this.shadowOffsetY = 0;
        _this.shadowColor = Color4.Black();
        _this.hTextAlign = 'left';
        _this.vTextAlign = 'bottom';
        _this.paddingTop = 0;
        _this.paddingRight = 0;
        _this.paddingBottom = 0;
        _this.paddingLeft = 0;
        return _this;
    }
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "outlineWidth", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIText.prototype, "outlineColor", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIText.prototype, "color", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "fontSize", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIText.prototype, "fontAutoSize", void 0);
    __decorate([
        ObservableComponent.component,
        __metadata("design:type", Font)
    ], UIText.prototype, "font", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIText.prototype, "value", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "lineSpacing", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "lineCount", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIText.prototype, "adaptWidth", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIText.prototype, "adaptHeight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIText.prototype, "textWrapping", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "shadowBlur", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "shadowOffsetX", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "shadowOffsetY", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIText.prototype, "shadowColor", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIText.prototype, "hTextAlign", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIText.prototype, "vTextAlign", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "paddingTop", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "paddingRight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "paddingBottom", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIText.prototype, "paddingLeft", void 0);
    UIText = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_TEXT_SHAPE)
    ], UIText);
    return UIText;
}(UIShape));
export { UIText };
/**
 * @public
 */
var UIInputText = /** @class */ (function (_super) {
    __extends(UIInputText, _super);
    function UIInputText(parent) {
        var _this = _super.call(this, parent) || this;
        _this.outlineWidth = 0;
        _this.outlineColor = Color4.Black();
        _this.color = Color4.Clear();
        _this.fontSize = 10;
        _this.value = '';
        _this.placeholder = '';
        _this.margin = 10;
        _this.hTextAlign = 'left';
        _this.vTextAlign = 'bottom';
        _this.focusedBackground = Color4.Black();
        _this.textWrapping = false;
        _this.shadowBlur = 0;
        _this.shadowOffsetX = 0;
        _this.shadowOffsetY = 0;
        _this.shadowColor = Color4.White();
        _this.paddingTop = 0;
        _this.paddingRight = 0;
        _this.paddingBottom = 0;
        _this.paddingLeft = 0;
        _this.onTextSubmit = null;
        _this.onChanged = null;
        _this.onFocus = null;
        _this.onBlur = null;
        _this.onTextChanged = new OnChanged(function (e) {
            var _a = e.value, value = _a.value, isSubmit = _a.isSubmit;
            // NOTE: here we want to keep the same `dirty` state as before changing `this.value`
            // because changing `this.value` will set the component as `dirty` and send a message to the renderer with it value
            // and that message is unnecesary (if the only thing that have changed is `this.value`) since that new value has come from the renderer itself
            var isDirty = _this.dirty;
            _this.value = value;
            _this.dirty = isDirty;
            if (isSubmit && _this.onTextSubmit) {
                var onSubmitValue = { text: value };
                _this.onTextSubmit.callback(onSubmitValue);
            }
            else if (!isSubmit && _this.onChanged) {
                var onChangeValue = {
                    value: value,
                    pointerId: e.pointerId
                };
                _this.onChanged.callback(onChangeValue);
            }
        });
        return _this;
    }
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "outlineWidth", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIInputText.prototype, "outlineColor", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIInputText.prototype, "color", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "fontSize", void 0);
    __decorate([
        ObservableComponent.component,
        __metadata("design:type", Font)
    ], UIInputText.prototype, "font", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIInputText.prototype, "value", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIInputText.prototype, "placeholder", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "margin", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIInputText.prototype, "hTextAlign", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", String)
    ], UIInputText.prototype, "vTextAlign", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIInputText.prototype, "focusedBackground", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIInputText.prototype, "textWrapping", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "shadowBlur", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "shadowOffsetX", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "shadowOffsetY", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIInputText.prototype, "shadowColor", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "paddingTop", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "paddingRight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "paddingBottom", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIInputText.prototype, "paddingLeft", void 0);
    __decorate([
        OnUUIDEvent.uuidEvent,
        __metadata("design:type", OnChanged)
    ], UIInputText.prototype, "onTextChanged", void 0);
    __decorate([
        OnUUIDEvent.uuidEvent,
        __metadata("design:type", Object)
    ], UIInputText.prototype, "onFocus", void 0);
    __decorate([
        OnUUIDEvent.uuidEvent,
        __metadata("design:type", Object)
    ], UIInputText.prototype, "onBlur", void 0);
    UIInputText = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_INPUT_TEXT_SHAPE),
        __metadata("design:paramtypes", [Object])
    ], UIInputText);
    return UIInputText;
}(UIShape));
export { UIInputText };
/**
 * @public
 */
var UIImage = /** @class */ (function (_super) {
    __extends(UIImage, _super);
    function UIImage(parent, source) {
        var _this = _super.call(this, parent) || this;
        _this.sourceLeft = 0;
        _this.sourceTop = 0;
        _this.sourceWidth = 1;
        _this.sourceHeight = 1;
        _this.paddingTop = 0;
        _this.paddingRight = 0;
        _this.paddingBottom = 0;
        _this.paddingLeft = 0;
        _this.sizeInPixels = true;
        _this.onClick = null;
        _this.source = source;
        return _this;
    }
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIImage.prototype, "sourceLeft", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIImage.prototype, "sourceTop", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIImage.prototype, "sourceWidth", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIImage.prototype, "sourceHeight", void 0);
    __decorate([
        ObservableComponent.component,
        __metadata("design:type", Object)
    ], UIImage.prototype, "source", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIImage.prototype, "paddingTop", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIImage.prototype, "paddingRight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIImage.prototype, "paddingBottom", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIImage.prototype, "paddingLeft", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIImage.prototype, "sizeInPixels", void 0);
    __decorate([
        OnUUIDEvent.uuidEvent,
        __metadata("design:type", Object)
    ], UIImage.prototype, "onClick", void 0);
    UIImage = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_IMAGE_SHAPE),
        __metadata("design:paramtypes", [UIShape, Object])
    ], UIImage);
    return UIImage;
}(UIShape));
export { UIImage };
/**
 * @public
 */
var UIScrollRect = /** @class */ (function (_super) {
    __extends(UIScrollRect, _super);
    function UIScrollRect() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.valueX = 0;
        _this.valueY = 0;
        _this.backgroundColor = Color4.Clear();
        _this.isHorizontal = false;
        _this.isVertical = false;
        _this.paddingTop = 0;
        _this.paddingRight = 0;
        _this.paddingBottom = 0;
        _this.paddingLeft = 0;
        _this.onChanged = null;
        return _this;
    }
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIScrollRect.prototype, "valueX", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIScrollRect.prototype, "valueY", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Color4)
    ], UIScrollRect.prototype, "backgroundColor", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIScrollRect.prototype, "isHorizontal", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Boolean)
    ], UIScrollRect.prototype, "isVertical", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIScrollRect.prototype, "paddingTop", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIScrollRect.prototype, "paddingRight", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIScrollRect.prototype, "paddingBottom", void 0);
    __decorate([
        ObservableComponent.field,
        __metadata("design:type", Number)
    ], UIScrollRect.prototype, "paddingLeft", void 0);
    __decorate([
        OnUUIDEvent.uuidEvent,
        __metadata("design:type", Object)
    ], UIScrollRect.prototype, "onChanged", void 0);
    UIScrollRect = __decorate([
        DisposableComponent('engine.shape', CLASS_ID.UI_SLIDER_SHAPE)
    ], UIScrollRect);
    return UIScrollRect;
}(UIShape));
export { UIScrollRect };
